src/cert.rs

```diff
@@ -221,9 +237,147 @@ fn remember_cert_extension<'a>(
+
+    fn next(&mut self) -> Option<Self::Item> {
+        (!self.reader.at_end()).then(|| CrlDistributionPoint::from_der(&mut self.reader))
+    }
+}
+
+/// A certificate revocation list (CRL) distribution point, describing a source of
+/// CRL information for a given certificate as described in RFC 5280 section 4.2.3.13[^1].
+///
+/// [^1]: <https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.13>
+pub(crate) struct CrlDistributionPoint<'a> {
+    /// distributionPoint describes the location of CRL information.
+    distribution_point: Option<untrusted::Input<'a>>,
+
+    /// reasons holds a bit flag set of certificate revocation reasons associated with the
+    /// CRL distribution point.
+    pub(crate) reasons: Option<der::BitStringFlags<'a>>,
+
+    /// when the CRL issuer is not the certificate issuer, crl_issuer identifies the issuer of the
+    /// CRL.
+    pub(crate) crl_issuer: Option<untrusted::Input<'a>>,
+}
+
+impl<'a> CrlDistributionPoint<'a> {
+    fn from_der(der: &mut untrusted::Reader<'a>) -> Result<Self, Error> {
+        // RFC 5280 section §4.2.1.13:
+        //   A DistributionPoint consists of three fields, each of which is optional:
+        //   distributionPoint, reasons, and cRLIssuer.
+        let mut result = CrlDistributionPoint {
+            distribution_point: None,
+            reasons: None,
+            crl_issuer: None,
+        };
+
+        der::nested(der, Tag::Sequence, Error::BadDer, |der| {
+            const DISTRIBUTION_POINT_TAG: u8 = CONTEXT_SPECIFIC | CONSTRUCTED;
+            const REASONS_TAG: u8 = CONTEXT_SPECIFIC | 1;
+            const CRL_ISSUER_TAG: u8 = CONTEXT_SPECIFIC | CONSTRUCTED | 2;
+
+            while !der.at_end() {
+                let (tag, value) = der::read_tag_and_get_value(der)?;
+                match tag {
+                    DISTRIBUTION_POINT_TAG => {
+                        set_extension_once(&mut result.distribution_point, || Ok(value))?
+                    }
+                    REASONS_TAG => {
+                        set_extension_once(&mut result.reasons, || der::bit_string_flags(value))?
+                    }
+                    CRL_ISSUER_TAG => set_extension_once(&mut result.crl_issuer, || Ok(value))?,
+                    _ => return Err(Error::BadDer),
+                }
+            }
+
+            // RFC 5280 section §4.2.1.13:
+            //   a DistributionPoint MUST NOT consist of only the reasons field; either distributionPoint or
+            //   cRLIssuer MUST be present.
+            if result.distribution_point.is_none() && result.crl_issuer.is_none() {
+                return Err(Error::MalformedExtensions);
+            }
+
+            Ok(result)
+        })
+    }
+
+    /// Return the distribution point names (if any).
+    #[allow(dead_code)] // TODO(@cpu): remove this once used in CRL validation.
+    pub(crate) fn names(&self) -> Result<Option<DistributionPointName<'a>>, Error> {
+        match self.distribution_point {
```

Comment:
Style: consider hoisting the `Ok`, or maybe there's a `map().transpose()` contraption that could work here?

Reply:
Good call, thanks! I need to get better at spotting this pattern :sweat_smile: